File: web_pra/web7/models/todo.py
from models import Model


class Todo(Model):
    def __init__(self, form):
        # form 值的是 数据库里面存取的数据, 读出以后格式为字典
        super().__init__()
        self.id = form.get('id', None)
        self.title = form.get('title', '')
        self.user_id = form.get('user_id', -1)

    @classmethod
    def update(cls, id, title):
        todos = cls.all()
        for t in todos:
            if t.id == id:
                t.title = title
        cls.save(todos)

    @classmethod
    def remove(cls, id):
        """
        删除一个 todo 的时候 其他的 todo 的 id
        """
        todos = cls.all()
        for t in todos:
            if t.id == id:
                todos.remove(t)
                break
        cls.save(todos)

    @classmethod
    def add(cls, todo, user_id):
        """
        增加一个 todo
        """
        todos = cls.all()
        # 新加的 id 需要重设 id
        todo.id = len(todos) + 1
        todo.user_id = user_id
        todos.append(todo)
        cls.save(todos)

File: 微博程序/ajax版/static/todo.js
const todoTemplate = (todo) => {
    const title = todo.title
    const id = todo.id
    const t = `
        <div class="todo-cell" id='todo-${id}' data-id="${id}">
            <span class='todo-title'>${title}</span>
            <button class="todo-edit">编辑</button>
            <button class="todo-delete">删除</button>
        </div>
        `
    return t
}

const insertEditForm = (cell) => {
    const form = `
        <div class='todo-edit-form'>
            <input class="todo-edit-input">
            <button class='todo-update'>更新</button>
        </div>
    `
    cell.insertAdjacentHTML('beforeend', form)
}


class Todo {
    constructor() {
        this.todoListDiv = e('.todo-list')
        this.init()
    }

    init() {
        this.loadTodos()
        this.bindEvents()
    }

    insertTodo(todo) {
        // 用来插入一条 todo 数据
        const todoCell = todoTemplate(todo)
        appendHtml(this.todoListDiv, todoCell)
    }

    bindEvents() {
        this.bindEventAdd()
        this.bindEventDelete()
        this.bindEventEdit()
        this.bindEventUpdate()
    }

    loadTodos() {
        TodoApi.all(todos => {
            for (let i = 0; i < todos.length; i++) {
                const todo = todos[i]
                this.insertTodo(todo)
            }
        })
    }

    bindEventAdd() {
        const addBtn = e('#id-button-add')
        bindEvent(addBtn, 'click', event => {
            const input = e('#id-input-todo')
            const title = input.value
            const form = {
                title: title,
            }
            TodoApi.add(form, todo => {
                this.insertTodo(todo)
            })
        })
    }

    bindEventDelete() {
        // 事件委托
        bindEvent(this.todoListDiv, 'click', event => {
            const target = event.target
            if (target.classList.contains('todo-delete')) {
                const todoCell = target.parentElement
                const todo_id = todoCell.dataset.id
                TodoApi.delete(todo_id, data => {
                    console.log('删除的数据为: ', data)
                    todoCell.remove()
                })
            }
        })
    }

    bindEventEdit() {
        bindEvent(this.todoListDiv, 'click', event => {
            const target = event.target
            if (target.classList.contains('todo-edit')) {
                const todoCell = target.parentElement
                insertEditForm(todoCell)
            }
        })
    }

    bindEventUpdate() {
        bindEvent(this.todoListDiv, 'click', event => {
            const target = event.target
            if (target.classList.contains('todo-update')) {
                const editForm = target.parentElement
                const title = editForm.querySelector('.todo-edit-input').value
                const todoCell = target.closest('.todo-cell')
                const todo_id = todoCell.dataset.id
                const form = {
                    id: todo_id,
                    title: title,
                }
                TodoApi.update(form, todo => {
                    const titleSpan = todoCell.querySelector('.todo-title')
                    titleSpan.innerHTML = todo.title
                })
            }
        })
    }
}


const __main = () => {
    todoController = new Todo()
}

__main()